blake2/src/blake2b.rs

```diff
@@ -1,18 +1,295 @@
+pub const PERSONALBYTES: usize = 2 * size_of::<Word>();
+
+/// The number input bytes passed to each call to the compression function. Small benchmarks need
+/// to use an even multiple of `BLOCKBYTES`, or else their apparent throughput will be low.
+pub const BLOCKBYTES: usize = 16 * size_of::<Word>();
+
+const IV: [Word; 8] = [
+    0x6A09E667F3BCC908,
+    0xBB67AE8584CAA73B,
+    0x3C6EF372FE94F82B,
+    0xA54FF53A5F1D36F1,
+    0x510E527FADE682D1,
+    0x9B05688C2B3E6C1F,
+    0x1F83D9ABFB41BD6B,
+    0x5BE0CD19137E2179,
+];
+
+const SIGMA: [[u8; 16]; 12] = [
+    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
+    [14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3],
+    [11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4],
+    [7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8],
+    [9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13],
+    [2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9],
+    [12, 5, 1, 15, 14, 13, 4, 10, 0, 7, 6, 3, 9, 2, 8, 11],
+    [13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10],
+    [6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5],
+    [10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0],
+    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
+    [14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3],
+];
+
+/// Compute the BLAKE2b hash of a slice of bytes all at once, using default
+/// parameters.
+///
+/// # Example
+///
+/// ```
+/// # use blake2::blake2b::{blake2b, Params};
+/// let expected = "ca002330e69d3e6b84a46a56a6533fd79d51d97a3bb7cad6c2ff43b354185d6d\
+///                 c1e723fb3db4ae0737e120378424c714bb982d9dc5bbd7a0ab318240ddd18f8d";
+/// let hash = blake2b(b"foo");
+/// assert_eq!(expected, &hash.to_hex());
+/// ```
+pub fn blake2b(input: &[u8]) -> Hash {
+    Params::new().hash(input)
+}
+
+/// Blake2b instance with a fixed output.
+#[derive(Clone, Default)]
+pub struct Blake2b {
+    params: Params,
+    state: State,
+}
+
+impl Blake2b {
+    /// Creates a new hashing context with the full set of sequential-mode parameters.
+    pub fn with_params(key: &[u8], salt: &[u8], persona: &[u8]) -> Self {
+        Params::new()
+            .key(key)
+            .salt(salt)
+            .personal(persona)
+            .clone()
+            .into()
+    }
+}
+
+impl From<Params> for Blake2b {
+    fn from(params: Params) -> Self {
+        let state = params.to_state();
+        Self { params, state }
+    }
+}
+
+impl BlockInput for Blake2b {
+    type BlockSize = U128;
+}
+
+impl Update for Blake2b {
+    fn update(&mut self, data: impl AsRef<[u8]>) {
+        self.state.update(data.as_ref());
+    }
+}
+
+impl FixedOutputDirty for Blake2b {
+    type OutputSize = U64;
+
+    fn finalize_into_dirty(&mut self, out: &mut Output) {
+        let hash = self.state.finalize();
+        out.copy_from_slice(hash.as_bytes());
+    }
+}
+
+impl Reset for Blake2b {
+    fn reset(&mut self) {
+        self.state = self.params.to_state();
```

Comment:
This is probably the main questionable thing with the current implementation. Every digest instance is keeping both `Params` and `State` just to implement reset this way.

Adding a first-class `reset` method to `State` would address that problem.

Reply:
Yes that's doable, though I think the only way to implement it is to follow the same strategy one level down, stashing a `Params` object in `State` just in case.

There might be some exceptionally lean callers that regret adding ~64 bytes to the BLAKE2s struct, but...in practive I kind of doubt it.